src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -1503,31 +1505,45 @@ internal Type[] FunctionPointerReturnAndParameterTypes
                         return ConstructName(ref m_name, TypeNameFormatFlags.FormatBasic);
 
                     case TypeNameKind.FullName:
-                        // We exclude the types that contain generic parameters because their names cannot be roundtripped.
-                        // We allow generic type definitions (and their refs, ptrs, and arrays) because their names can be roundtriped.
-                        // Theoretically generic types instantiated with generic type definitions can be roundtripped, e.g. List`1<Dictionary`2>.
-                        // But these kind of types are useless, rare, and hard to identity. We would need to recursively examine all the
-                        // generic arguments with the same criteria. We will exclude them unless we see a real user scenario.
-                        if (!m_runtimeType.GetRootElementType().IsGenericTypeDefinition && m_runtimeType.ContainsGenericParameters)
-                            return null;
-
-                        // Exclude function pointer; it requires a grammar update and parsing support for Type.GetType() and friends.
-                        // See https://learn.microsoft.com/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names.
-                        if (m_runtimeType.IsFunctionPointer)
-                            return null;
-
-                        // No assembly.
-                        return ConstructName(ref m_fullname, TypeNameFormatFlags.FormatNamespace | TypeNameFormatFlags.FormatFullInst);
+                        return ConstructFullName(ref m_fullName);
 
                     case TypeNameKind.ToString:
                         // No full instantiation and assembly.
                         return ConstructName(ref m_toString, TypeNameFormatFlags.FormatNamespace);
 
+                    case TypeNameKind.AssemblyQualifiedName:
+                        string? typeFullName = null;
+                        // This function returns the byref argument, so the return can be ignored.
+                        ConstructFullName(ref typeFullName);
+                        Debug.Assert(typeFullName == GetName(TypeNameKind.FullName));
+                        if (typeFullName is null)
+                            return null;
+                        return m_assemblyQualifiedName ??= Assembly.CreateQualifiedName(m_runtimeType.Assembly.FullName, typeFullName);
```

Comment:
I assume that you have looked at the TypeNameFormatFlags and determined that it is better to create two strings and concatenate them instead of building the full string directly. Is there a simple explanation for why producing the fully qualified name directly is not a good idea?

Reply:
I will put up a commit with the changes. We can then decide which ones we want.

Reply:
I do not see where these two calls GetDisplayName can produce different value.

BTW: FormatNoVersion looks unused - it can be deleted.

Reply:
The `ASM_DISPLAYF_CONTENT_TYPE` does seem to be a nop in general, but `ASM_DISPLAYF_PUBLIC_KEY_TOKEN` does seem to add the public key token. Perhaps I'm not understanding what these things do.

Reply:
Nevermind... I missed this statement:

https://github.com/dotnet/runtime/blob/dc0033449c7d98a1593b1859573a6449d06c56c7/src/coreclr/vm/coreassemblyspec.cpp#L330-L333

Reply:
Edited the code wrong, sorry.

Reply:
Okay, so the only two options that are missing are:

```
    ASM_DISPLAYF_PROCESSORARCHITECTURE  = 0x20,
    ASM_DISPLAYF_RETARGET               = 0x80,
```

I've no idea what `ASM_DISPLAYF_RETARGET` does, but looking around it looks dead. I think `ASM_DISPLAYF_PROCESSORARCHITECTURE` is still valid though.

Reply:
I think I'm just going to change the `FormatAssembly` to mean `ASM_DISPLAYF_FULL`. This adds:

```
    ASM_DISPLAYF_PROCESSORARCHITECTURE  = 0x20,
    ASM_DISPLAYF_RETARGET               = 0x80,
```